### PR TITLE
Improve performance of MarkSeries using shouldComponentUpdate

### PR DESCRIPTION
**Pull request by jonsadka**
## Prompt
Improve rerendering performance of `MarkSeries` component by using `shouldComponentUpdate`

## Problem
The `MarkSeries` component receives several props of type function (`onNearestX`, `onNearestXY`, etc.). If these functions are called, the `MarkSeries` component will automatically rerender all datapoints by mapping `this._renderCircle` on them, regardless if the change is necessary.

## Proposal
Only rerender when props affecting the view are changed. 

### Before
kapture 2017-11-13 at 11 29 13-min

### After
kapture 2017-11-13 at 10 48 02-min

### CODE REVIEW

File: src/plot/series/mark-series.js [hunk 1]
```diff
@@ -32,6 +33,35 @@ const predefinedClassName = 'rv-xy-plot__series rv-xy-plot__series--mark';
...
+      [this.props.colorDomain, nextProps.colorDomain],
+      [this.props.colorRange, nextProps.colorRange],
+      [this.props.data, nextProps.data],
+      [this.props.fillDomain, nextProps.fillDomain],
+      [this.props.fillRange, nextProps.fillRange],
+      [this.props.innerHeight, nextProps.innerHeight],
+      [this.props.innerWidth, nextProps.innerWidth],
+      [this.props.marginLeft, nextProps.marginLeft],
+      [this.props.marginTop, nextProps.marginTop],
+      [this.props.nullAccessor, nextProps.nullAccessor],
```
Comment (jckr) [inline src/plot/series/mark-series.js]: what is this?

Comment (jonsadka) [reply src/plot/series/mark-series.js]: javascripts

Comment (jckr) [reply src/plot/series/mark-series.js]: specifically - nullAccessor? where does that come from?

Comment (jonsadka) [reply src/plot/series/mark-series.js]: Oh. That's an optional prop which returns a boolean but for `MarkSeries` its defined at the very bottom of the file 
```javascript
MarkSeries.defaultProps = {
  nullAccessor: () => true
};
```
and from the docs
```markdown
#### nullAccessor (optional)
Type: `function`  
Default: `null`  
A function that will be invoked for each data element that will return a boolean that specifies if the data point should be drawn or not. For more information see [the D3 documentation](https://github.com/d3/d3-shape#line_defined).
```

File: showcase/plot/scatterplot-large.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  XYPlot,
+  YAxis
+} from 'index';
+
+const FlexibleXYPlot = makeWidthFlexible(XYPlot);
+
+export default class Example extends React.Component {
+  componentWillMount() {
+    this.setState({
+      data: randomDatapoints(randomInt(700)),
```
Comment (ericsoco) [inline showcase/plot/scatterplot-large.js]: Why are you randomizing the number of points to render? Why not let that be a constant (but still randomize each datapoint)? Also, seems like a "scatterplot-large" would be significantly more than 700 points...I would expect a viz framework to be able to gracefully handle at least a couple thousand.

File: src/plot/series/mark-series.js [hunk 3]
```diff
@@ -32,6 +33,44 @@ const predefinedClassName = 'rv-xy-plot__series rv-xy-plot__series--mark';
 const DEFAULT_STROKE_WIDTH = 1;
 
 class MarkSeries extends AbstractSeries {
+  shouldComponentUpdate(nextProps, nextState) {
+    const isMissingDataProp = !this.props.data || !nextProps.data;
+    const isMissingNullAccessorProp = !this.props.nullAccessor || !nextProps.nullAccessor;
+    return [
```
Comment (ericsoco) [inline src/plot/series/mark-series.js]: There's no reason to recreate this array every time this func is called. Make a const array outside the class declaration with just the string prop names, and loop over that and pull values out of `this.props` and `nextProps` for each string name. More legible and a slight optimization.

File: src/plot/series/mark-series.js [hunk 4]
```diff
@@ -32,6 +33,44 @@ const predefinedClassName = 'rv-xy-plot__series rv-xy-plot__series--mark';
...
+      [this.props.strokeDomain, nextProps.strokeDomain],
+      [this.props.strokeRange, nextProps.strokeRange],
+      [this.props.strokeWidth, nextProps.strokeWidth],
+      [this.props.style, nextProps.style],
+      [this.props.xDomain, nextProps.xDomain],
+      [this.props.xRange, nextProps.xRange],
+      [this.props.yDomain, nextProps.yDomain],
+      [this.props.yRange, nextProps.yRange]
+    ].some(([a, b]) => !shallowequal(a, b)) || (
+      isMissingDataProp ?
```
Comment (ericsoco) [inline src/plot/series/mark-series.js]: I'd pull this out into its own function for two reasons:
1. legibility: you can name + comment the function to describe this comparison you're doing
2. reusability: seems like a util that could be useful elsewhere.

Also, this diff should definitely include unit tests for this logic.

**Comment by ericsoco** [comment]
Is this optimization useful in other series?

And can you provide the benchmarks you used to validate it? (I know you did some testing, that should be in this PR for future reference.)

I'm generally a bit wary of `shouldComponentUpdate` optimizations that are not super simple, they often seem to backfire on you...and React DOM diffing often is all the optimization you need given that rendering is usually the most expensive part.

**Comment by mcnuttandrew** [comment]
I would strongly caution against any work that reproduces every single one of the props that a series might take in a single place. This effectively means that anything a new prop is introduced it will need to be duplicated into this one specific series. 

I agree with @ericsoco's sentiment that this could be moved into abstract series so that components are not updated when their listeners are called. Given that we name listeners in a standard way (onNearestX onMouseOver onBlanketyBlank) it could probably be inferred which props need to be checked from a simple heuristic regex. Adding this to AbstractSeries would certainly be a big win and would probably lighten up the load of some of the other more complex series, such as contour etc.

Also plus 1 for some bench marks, proving you're right is good.

**Pull request closed without merging**